Add QAT support for distributed finetuning

andrewor14 opened the pull request:
**Summary:** This commit adds the option to run quantization-aware training (QAT) during finetuning. QAT refers to "fake quantizing" the weights and activations during training, which performs the following transformation on the inputs but still keeps all intermediate values in floating point:

```
x_q = clamp((x_bf16 / scale) + zp)
x_fq = (x_q - zp) * scale
```

Currently only 8-bit per token dynamic activations + 4-bit grouped per channel weights (8da4w) is supported. Users can enable this by specifying a QAT quantizer in their config files:

```
tune run --nnodes 1 --nproc_per_node 8 qat_distributed --config llama3/8B_qat_full

# or add this to your config file
# quantizer:
#   _component_: torchtune.utils.quantization.Int8DynActInt4WeightQATQuantizer
#   groupsize: 256
```

**Test Plan:**

Initial results for Llama2 demonstrate that QAT is able to recover the loss in accuracy from quantization by about half for some tasks (last two rows):

|                 | hellaswag |          |     wikitext    |                 |               | arc_easy |          | arc_challenge |          |
|-----------------|:---------:|----------|:---------------:|-----------------|---------------|:--------:|----------|:-------------:|----------|
|                 |    acc    | acc_norm | word_perplexity | byte_perplexity | bits_per_byte |    acc   | acc_norm |      acc      | acc_norm |
| No quant |  59.659%  |  76.927% |      12.183     |      1.596      |     0.674     |  76.010% |  72.054% |    48.720%    |  47.867% |
| PTQ             |  57.150%  |  74.945% |      12.995     |      1.615      |     0.692     |  75.968% |  70.118% |    46.416%    |  45.904% ||
| QAT |  58.504%  |  76.170% |      12.199     |      1.596      |     0.675     |  76.431% |  71.928% |    47.184%    |  48.123% |
| PTQ degradation |  -2.509%  |  -1.982% |      +0.812      |      +0.019      |     +0.018     |  -0.042% |  -1.936% |    -2.304%    |  -1.963% |
| QAT degradation |  -1.155%  |  -0.757% |      +0.016      |      +0.000      |     +0.001     |  0.421%  |  -0.126% |    -1.536%    |  0.256%  |

andrewor14 commented:
By the way, the tests are failing due to torchtune's dependency on an old version of torchao, which doesn't have QAT support yet. We're about to release 0.2.0 on the torchao side (aiming early next week), so the tests won't pass until then.

kartikayk reviewed (commented):
Thanks for adding this @andrewor14! A few high-level comments apart from the one on teasing this out into a separate recipe:

> Currently only 8-bit per token dynamic activations + 4-bit grouped per channel weights (8da4w) is supported

A couple of questions:
- Do we plan on adding other quantization methods? Or whats the long term support plan for this?
- We need a lot of documentation here to make sure users understand what this actually means. Is this reasonably well understood by the general audience? for example, I'm not sure about all of the details here.

We also need to consider how to lower the bar for adopotion here. I think this needs a tutorial or a deepdive added to the torchtune docs.

cc: @ebsmothers

andrewor14 commented on recipes/full_finetune_distributed.py:
```diff
@@ -116,6 +116,8 @@ def __init__(self, cfg: DictConfig) -> None:
         # Training cfg
         self._resume_from_checkpoint = cfg.resume_from_checkpoint
         self._gradient_accumulation_steps = cfg.gradient_accumulation_steps
+        self._qat_enable_fake_quant_step = cfg.get("qat_enable_fake_quant_step", None)
```
This config is about when to enable fake quant, not whether to enable QAT itself. E.g. setting this to 1000 means we will run regular finetuning for the first 1000 steps, and only enable fake quant after 1000 steps. I'll add better comments/docs about this. Do you think the name makes sense or do you have suggestions?

kartikayk replied:
Ah, then I grossly misunderstood :) Maybe something like ```quant_after_n_steps```?

andrewor14 commented:
Hi @kartikayk, thanks for the comments, responding inline:

> * Do we plan on adding other quantization methods? Or whats the long term support plan for this?
Yes, in the long term we do plan to support other QAT configurations (e.g. 2- or 3-bit weight only if we can get good results), that's why I kept the quantizer specification general.

> * We need a lot of documentation here to make sure users understand what this actually means. Is this reasonably well understood by the general audience? for example, I'm not sure about all of the details here. We also need to consider how to lower the bar for adopotion here. I think this needs a tutorial or a deepdive added to the torchtune docs.

For sure. Should I add the README in this PR and add the tutorial separately?

andrewor14 commented:
Hi @kartikayk, thanks for the comments, responding inline:

> * Do we plan on adding other quantization methods? Or whats the long term support plan for this?

Yes, in the long term we do plan to support other QAT configurations (e.g. 2- or 3-bit weight only if we can get good results), that's why I kept the quantizer specification general.

> * We need a lot of documentation here to make sure users understand what this actually means. Is this reasonably well understood by the general audience? for example, I'm not sure about all of the details here. We also need to consider how to lower the bar for adopotion here. I think this needs a tutorial or a deepdive added to the torchtune docs.

For sure. Should I add the README in this PR and add the tutorial separately?

kartikayk commented:
> Should I add the README in this PR and add the tutorial separately?

I think a tutorial/deep-dive in the docs would be really helpful. You can add some details on what the quantization methods mean as well - I think this will be very useful and make the flow more noob friendly.

Here are some pointers:
- Example PR with a tutorial (checkout the e2e_flow.rst file): https://github.com/pytorch/torchtune/pull/690/files and this with the LoRA tutorial: https://github.com/pytorch/torchtune/pull/368
- Working with the live docs in torchtune: https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md#building-docs

I don't mind this as a follow up. 

@ebsmothers let me know if you have differing thoughts on this.

kartikayk commented on recipes/full_finetune_distributed.py:
```diff
@@ -288,6 +292,18 @@ def _setup_model(
                 ac_option,
             )
 
+        # Optionally apply quantization-aware training during finetuning
```
We've been pretty good at making sure we update all of the recipes with new features. I do think QAT is something we want to publicize heavily and so having its own recipe opens up avenues for future work as well

ebsmothers commented:
Thanks for the PR! I think I'm in agreement with most everything that's been said already: (1) separate recipe for QAT (I agree this will make it easier to scale once we add other quant techniques anyways) and (2) add some kind of tutorial but as a follow-up (also happy to provide any pointers or guidance you need in advance).

andrewor14 commented on tests/recipes/test_qat_distributed.py:
```diff
@@ -0,0 +1,102 @@
+        runpy.run_path(TUNE_PATH, run_name="__main__")
+        loss_values = get_loss_values_from_metric_logger(log_file)
+        # TODO: delete
+        print("LOSS VALUES ARE ", loss_values)
```
note to self: delete

ebsmothers commented on recipes/qat_distributed.py:
```diff
@@ -0,0 +1,644 @@
+                    running_loss = 0
+                    num_tokens = 0
+                    t0 = time.perf_counter()
+                global_step += 1
```
Are we double-incrementing global_step between here and L573?

andrewor14 replied:
oh great, I didn't realize we already have a `self.global_step`. Will remove this

ebsmothers commented on recipes/quantization.md:
````diff
@@ -1,13 +1,43 @@
+dtype to a lower precision.
+
+```
+tune run qat_distributed --config qat
````
Need nproc_per_node here? Also nit: Add some segue before the code block, e.g. "QAT can be run via the command"

ebsmothers commented on recipes/quantization.md:
```diff
@@ -1,13 +1,43 @@
+This produces an unquantized model in the original data type. To get an actual quantized model,
+follow this with `tune run quantize` while specifying the same quantizer in the config.
+
+The `qat_distributed` recipe is also compatible with any full finetuning config,
+as long as a QAT quantizer is specified in the config:
```
Sort of. But strictly speaking we associate a given config with only a single recipe. In that case we technically don't support something like e.g.

```
tune run --nproc_per_node 4 qat_distributed --config mistral/7B_full \
quantizer=torchtune.utils.quantization.Int8DynActInt4WeightQATQuantizer \
quantizer.groupsize=256
```

(This is kinda related to my other comment about how we want to actually expose QAT configs)

ebsmothers commented on recipes/quantization.md:
````diff
@@ -1,13 +1,43 @@
+  groupsize: 256
+```
+
+Currently only `torchtune.utils.quantization.Int8DynActInt4WeightQATQuantizer`
+is supported. This refers to int8 dynamic per token activation quantization
+combined with int4 grouped per axis weight quantization.
````
nit: can add a link to a reference or something here

andrewor14 replied:
We don't have a README or proper docs in torchao yet, will just link to the implementation for now

ebsmothers commented on recipes/quantize.py:
```diff
@@ -54,6 +54,15 @@ class QuantizationRecipe:
+        Args:
+            `groupsize` (int): a parameter of int4 weight only quantization,
+            it refers to the size of quantization groups which get independent quantization parameters
+            e.g. 32, 64, 128, 256, smaller numbers means more fine grained and higher accuracy
```
Presumably also higher memory, right?

ebsmothers commented on recipes/quantize.py:
```diff
@@ -93,7 +104,10 @@ def _setup_model(
     def quantize(self, cfg: DictConfig):
         t0 = time.perf_counter()
-        self._model = self._quantizer.quantize(self._model)
+        if "qat" in self._quantization_mode:
+            self._model = self._quantizer.convert(self._model)
+        else:
+            self._model = self._quantizer.quantize(self._model)
```
So we need to gate on quantization mode because the QAT checkpoints are in bf16, right? Noob question but if we are not doing any subsequent training why can't we just call `.quantize` directly and infer all the quantizer params from the checkpoint?

andrewor14 replied:
Technically we can, the numerics may be the the same, but officially the torchao QAT flow is:

```
quantizer = 8da4wQATQuantizer()
model = quantizer.prepare(model)
train(model)
model = quantizer.convert(model)
```

If we just call quantize here we would have to introduce a different quantizer

```
quantizer = 8da4wQATQuantizer()
model = quantizer.prepare(model)
train(model)
ptq_quantizer = 8da4wQuantizer()
ptq_quantizer.quantize(model)
```

I feel it's better to call the complete QAT flow rather than to switch quantizers in the middle

andrewor14 commented on pyproject.toml:
```diff
@@ -25,7 +25,8 @@ dependencies = [
     # Quantization
-    "torchao==0.1",
+    # TODO: update to 0.3
+    "torchao==0.2",
```
I think we'll wait till 0.3 is available

andrewor14 commented on recipes/configs/qat.yaml:
```diff
@@ -0,0 +1,78 @@
+# Config for multi-device QAT finetuning in qat_distributed.py
```
Sounds good. I mostly just followed `quantization.yaml`, `eleuther_evaluation.yaml`, and `generation.yaml` so far in this PR. I think these configs have the same problem. Both (2) and (3) make sense to me. What do you think?

andrewor14 replied:
Ok, I did (2) for now (put it in respective llama2 and llama3 dirs). Let me know if this sounds reasonable to you

ebsmothers replied:
Yeah I think this is reasonable. It is a bit of a pain to override all the necessary config fields to change models from the command line, so I think it makes sense to separate out QAT configs by model (since for (3) it'd get really verbose to change the model from the default config anyways). The downside is the configs are slightly less visible, but given we have a standalone top-level recipe this is OK imo.

andrewor14 commented:
Update: I think I've addressed all the comments and this PR is ready from my side. Do you have other comments @ebsmothers @kartikayk?

Note that landing is blocked right now on the torchao 0.3 release (currently scheduled for 6/26). This is because QAT was only added in torchao 0.2, but the following error was not fixed until torchao 0.3, so there's no other way to get the QAT feature unless we want to rely on nightlies, which we don't.

```
  File "/__w/_temp/conda_environment_9553124795/lib/python3.8/site-packages/torchtune/modules/common_utils.py", line 12, in <module>
    from torchao.dtypes.nf4tensor import NF4Tensor
  File "/__w/_temp/conda_environment_9553124795/lib/python3.8/site-packages/torchao/__init__.py", line 14, in <module>
    from . import _C
ImportError: /__w/_temp/conda_environment_9553124795/lib/python3.8/site-packages/torchao/_C.cpython-38-x86_64-linux-gnu.so: undefined symbol: _ZN5torch3jit11parseSchemaERKSs
```
